runtime/android/core/src/org/xwalk/core/XWalkExpansionContext.java

```diff
@@ -0,0 +1,312 @@
+        }
+
+        // Look what we have. We need to rename the expansion file
+        // to .apk for being able to load it with the DexClassLoader
+        String expPathApk = null;
+        String expPathObb = null;
+        for (String expName : expDir.list()) {
+            if (expName.startsWith(expType) && expName.endsWith(".apk")) {
+                expPathApk = expDir + "/" + expName;
+            } else if (expName.startsWith(expType) && expName.endsWith(".obb")) {
+                expPathObb = expDir + "/" + expName;
+            }
+        }
+
+        if (expPathApk == null && expPathObb == null) {
+            Log.e(TAG, "No expansion file found in " + expDir.getAbsolutePath());
+            return null;
+        }
+
+        File expFile = null;
+        if (getExpansionFileVersion(expPathObb) > getExpansionFileVersion(expPathApk)) {
+            expFile = new File(expPathObb);
+        } else if (expPathApk != null) {
+            expFile = new File(expPathApk);
+        }
+
+        if (expFile != null) {
+            Log.d(TAG, "Found expansion file " + expFile.getAbsolutePath());
+        } else {
+            Log.e(TAG, "No expansion file found in " + expDir.getAbsolutePath());
+        }
+
+        return expFile;
+    }
+
+    /**
+     * Extract LIBXWALKCORE and create dex class loader
+     * @param expFile Expansion file
+     * @return Expansion file object on success, otherwise null.
+     */
+    File prepareExpansionContent(File expFile) {
+
+        File expFileApk = new File(expFile.getAbsolutePath() + ".apk");
+        if (!expFile.renameTo(expFileApk)) {
+            Log.e(TAG, "Failed renaming expansion file to " + expFileApk.getAbsolutePath());
+            return null;
+        }
+
+        boolean extracted = extractLib(expFileApk);
+        if (!extracted) {
+            return null;
+        }
+
+        Log.d(TAG, "Prepared expansion content");
+        return expFileApk;
+    }
+
+    /**
+     * Extract LIBXWALKCORE from expansion file, so it can be JNI'd.
+     * @param expFile Expansion file
+     * @return true on success.
+     */
+    private boolean extractLib(File expFile) {
+
+        boolean haveLib = false;
+        for (String abi : Build.SUPPORTED_ABIS) {
+            String entryName = "lib/" + abi + "/" + LIBXWALKCORE;
+            haveLib = extractEntry(expFile, entryName);
+            if (haveLib)
+                break;
+        }
+
+        if (!haveLib) {
+            Log.e(TAG, "Failed to extract " + LIBXWALKCORE + " with supported ABI");
+            return false;
+        }
+
+        Log.d(TAG, "Extracted " + LIBXWALKCORE);
+        return true;
+    }
+
+    /**
+     * Extract entry from the expansion file.
+     * @param expFile Expansion file
+     * @param entryName Name of entry
+     * @return true on success.
+     */
+    private boolean extractEntry(File expFile, String entryName) {
+
+        JarFile jar;
+        try {
+            jar = new JarFile(expFile);
```

Comment:
Why not just use ZipFile to extract the entry as the expFile is actually a Zip file?